### PR TITLE
Add support for PostgreSQL Transport

### PR DESCRIPTION
**Pull request by jpalac**
Implementation of https://github.com/Particular/ServiceControl/issues/4254

Fixes https://github.com/Particular/ServiceControl/issues/4468

### CODE REVIEW

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            }
+
+            DisableDelayedDelivery(transport) = true;
+
+            transport.TransportTransactionMode = transport.GetSupportedTransactionModes().Contains(preferredTransactionMode) ? preferredTransactionMode : TransportTransactionMode.ReceiveOnly;
+
+            return transport;
+        }
+
+        public override string ToTransportQualifiedQueueName(string queueName) => $"public.{queueName}";
```
Comment (jpalac) [inline src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:75]: Thinking out loud, whether it's better with the public or the quotes. ie which one will pass more user mutations of the queue name. eg, if the function was something like:

```
string Delimiter = "\"";
string EscapedDelimiter = Delimiter + Delimiter;
if (queue.StartsWith(Delimiter) || queue.EndsWith(Delimiter))
{
    return queue;
}

return Delimiter + queue.Replace(Delimiter, EscapedDelimiter) + Delimiter;
```

File: src/ServiceControl.Transports.PostgreSql/DatabaseDetails.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            }
+            catch (NpgsqlException ex) when (IsConnectionOrLoginIssue(ex))
+            {
+                throw new Exception($"Could not connect to '{DatabaseName}' PostgreSQL database.", ex);
+            }
+        }
+
+        static bool IsConnectionOrLoginIssue(NpgsqlException x)
+        {
+            //TODO postgres - any other errors?
```
Comment (andreasohlund) [inline src/ServiceControl.Transports.PostgreSql/DatabaseDetails.cs:46]: Anything more we need to do here?

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    using NServiceBus.Logging;
+    using NServiceBus.Transport.PostgreSql;
+
+    public class PostgreSqlTransportCustomization : TransportCustomization<PostgreSqlTransport>
+    {
+        protected override void CustomizeTransportForPrimaryEndpoint(EndpointConfiguration endpointConfiguration, PostgreSqlTransport transportDefinition, TransportSettings transportSettings)
+        {
+            transportDefinition.TransportTransactionMode = TransportTransactionMode.SendsAtomicWithReceive;
+            //var routing = new RoutingSettings(endpointConfiguration.GetSettings());
+            //routing.EnableMessageDrivenPubSubCompatibilityMode();
```
[the remark this answers is not in the document]
Comment (jpalac) [reply src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:17]: Is there an equivalent of `EnableMessageDrivenPubSubCompatibilityMode `on the postgres transport? If not relevant then yes we can remove this.

File: src/ServiceControl.Transports.PostgreSql/DatabaseDetails.cs (hunk 2 above)
Comment (jpalac) [reply src/ServiceControl.Transports.PostgreSql/DatabaseDetails.cs]: I've just tested it, fixed it and removed the comment

**Comment by andreasohlund** [comment]
@jpalac @PhilBastian ready for final review

File: src/ServiceControl.Transports.PostgreSql/transport.manifest [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      "Name": "PostgreSQL",
+      "DisplayName": "PostgreSQL",
+      "AssemblyName": "ServiceControl.Transports.PostgreSql",
+      "TypeName": "ServiceControl.Transports.PostgreSql.PostgreSqlTransportCustomization, ServiceControl.Transports.PostgreSql",
+      "DefaultPrimaryMaximumConcurrencyLevel": 10,
+      "DefaultAuditMaximumConcurrencyLevel": 10,
+      "DefaultMonitoringMaximumConcurrencyLevel": 10,
+      "SampleConnectionString": "Server=<ServerName>;Database=nservicebus;Port=5432;User Id=<Username>;Password=<Password>;Queue Schema=myschema;Subscriptions Table=schema.tablename",
+      "AvailableInSCMU": true,
+      "Help": "Specify optional 'Queue Schema' to override the default schema. Specify optional 'Subscriptions Table' to override the default subscriptions table location. When integrated authentication is specified in the PostgreSql connection string the the current installing user is used to create the required SQL tables structure not the service account."
```
Comment (jpalac) [inline src/ServiceControl.Transports.PostgreSql/transport.manifest:13]:
```suggestion
      "Help": "Specify optional 'Queue Schema' to override the default schema. Specify optional 'Subscriptions Table' to override the default subscriptions table location."
```

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        if (subscriptionsTableSetting != null)
+        {
+            var subscriptionsAddress = QueueAddress.Parse(subscriptionsTableSetting);
+
+            subscriptions.SubscriptionTableName =
+                new SubscriptionTableName(subscriptionsAddress.Table,
+                    subscriptionsAddress.Schema ?? customSchema);
+        }
+
+        if (transportSettings.GetOrDefault<bool>("TransportSettings.EnableDtc"))
```
Comment (PhilBastian) [inline src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:55]: not originating from this PR I know, but why is this not a string constant somewhere?

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        DisableDelayedDelivery(transport) = true;
+
+        transport.TransportTransactionMode = transport.GetSupportedTransactionModes().Contains(preferredTransactionMode) ? preferredTransactionMode : TransportTransactionMode.ReceiveOnly;
+
+        return transport;
+    }
+
+    protected override string ToTransportQualifiedQueueNameCore(string queueName)
+    {
+        const string delimiter = "\"";
```
Comment (PhilBastian) [inline src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:69]: is this the same qualification scheme used in `PostgreSqlNameHelper`?

File: src/ServiceControl.Transports.PostgreSql/QueueTableSnapshot.cs [hunk 7]
```diff
@@ -0,0 +1,6 @@
+namespace ServiceControl.Transports.PostgreSql;
+
+public class BrokerQueueTableSnapshot(BrokerQueueTable details) : BrokerQueueTable(details.DatabaseDetails, details.QueueAddress)
+{
+    public long RowVersion { get; set; }
```
Comment (PhilBastian) [inline src/ServiceControl.Transports.PostgreSql/QueueTableSnapshot.cs:5]: RowVersion is still correct for Postgres?

**Review by PhilBastian — APPROVED** [review]

File: src/ServiceControl.Transports.PostgreSql/QueueTableSnapshot.cs (hunk 7 above)
Comment (jpalac) [reply src/ServiceControl.Transports.PostgreSql/QueueTableSnapshot.cs:5]: it's just an internal name for calculating the throughput

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs (hunk 6 above)
Comment (jpalac) [reply src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:69]: yes - are you thinking of making those public and using them here?

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs (hunk 5 above)
Comment (jpalac) [reply src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:55]: Well it's only used once so probably didn't see a point to it?

File: src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs (hunk 6 above)
Comment (PhilBastian) [reply src/ServiceControl.Transports.PostgreSql/PostgreSqlTransportCustomization.cs:69]: makes more sense than duplicating code

**Pull request merged**